## FastQC

From [FastQC
website](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/):
"FastQC aims to provide a simple way to do some quality control checks
on raw sequence data coming from high throughput sequencing
pipelines. It provides a modular set of analyses which you can use to
give a quick impression of whether your data has any problems of which
you should be aware before doing any further analysis. "

This is a useful program that performs some of the basic diagnostics
on a FASTQ sequence file, looking for common issues that could appear
in your libraries.

First we would download the FastQC program:

* If you are using a PC or Linux, you can download it
  [here](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip)
* Unzip the file, and open the FastQC folder
* Doubleclick on run_fastqc.bat
![PC batch file](screenshots/fastqc_pc.png | width=300)


* If you are using a Mac, you can download it [here](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.dmg)
* Doubleclick on the downloaded DMG file to open it
* Doubleclick on the FastQC app
![Mac app](screenshots/fastqc_mac.png | width=300)
